### ISSUE TITLE
Use ES6 module export/import 

### ISSUE THREAD

**Issue by jbucaran** [hyperapp/hyperapp#86]
# tl;dr
* Rollup/webpack can't tree-shake your hyperapp application.
> Tree-shaking is the process of only including the code that is used in your app. It is similar to dead code elimination. See [Tree-shaking vs Dead-Code Elimination](https://medium.com/@Rich_Harris/tree-shaking-versus-dead-code-elimination-d3765df85c80#.z25jhearl).
* Rollup/webpack can tree-shake if we use ES6 module import/export semantics.
* Using ES6 module import/export **does not** mean the project will be rewritten in ES6, or that you have to use import/export in your application.

---

## CommonJS (current)
For a small/medium-ish project, using commonjs equals less tooling. But _only_ when hacking directly on the source code of a library intended to run on node (e.g., not hyperapp). If you use `require("mylib")` you'd be requiring the compiled version of that library anyway!
  * Let's see how that'd work:
    1. `git clone .../mylib `
    2. Create a dummy.js file, `require("./")` and start hacking.
    3. Alternatively, start a node REPL and `require("./")`.
  * Notice I'm using the term _hacking_. Hacking, here, means contributing to the development of `mylib`, forking/rolling your own module based in `mylib`, etc.

You always need to bundle hyperapp with your app.js/index.js, because they are intended to run on a browser. Therefore, distributing hyperapp as a commonjs module offers no remarkable advantage to newcomers, hackers, etc.

## ES6 modules
Using ES6 import/export semantics is the [spec](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import), but what other advantage is there?

The answer is [tree-shaking](https://medium.com/@Rich_Harris/tree-shaking-versus-dead-code-elimination-d3765df85c80#.z25jhearl).

Because of the way index.js exports h, app and router, neither webpack or rollup is able to do tree shaking correctly on you app. In other words, even if you only `const { h } = require("hyperapp")`, your app's bundle still includes app+router, which we'd like to avoid and we **totally can** with tree-shaking.

If you were using browserify, you could be tree shaking via [rollupify](https://github.com/nolanlawson/rollupify) too!

> ⚠️ Webpack treeshaking is actually [broken](https://github.com/webpack/webpack/issues/2867), but that's likely to be fixed sooner than later.

### In the defense of commonjs
Actually, rollup can do treeshaking regardless, if we can just tweak index.js a bit.

Currently:
```js
module.exports = {
    router: require("./router"),
    app: require("./app"),
    h: require("./h")
}
```
But if we change it to:
```js
module.exports.h = require("./h")
module.exports.app = require("./app")
module.exports.router = require("./router")
```

Tree-shaking works.

**Comment by jbucaran** [comment]
The post above assumes we want index to export h, app and router. Alternatively, we could have index export only h and app and import router separately, for example:

```jsx
const { h, app } = require("hyperapp")
const router = require("hyperapp/router")
```

or

```jsx
import { h, app } from "hyperapp"
import router from "hyperapp/router"
```

**Comment by tunnckoCore** [comment]
1. One thing is that app should work without the router and router should work without the app. That's why router should be different package. It is just enough to make them understand each other.

> But if we change it to:

2. I believe it isn't needed, because rollup will just change module exports to `exports.default`. Another thing is that it does not activates tree-shaking on commonjs modules (understand, as it is currently, using rollup-plugin-node-resolve + rollup-plugin-commonjs combo). So no, tree-shaking won't work until we have import/export.

> The uncomfortable part is when building dist/hyperapp.js, rollup adds now a default property

3. Won't have. The umd bundle won't have such thing. As about the ES distributed build, this isn't a problem,because users will use ES6 too and will get what they need, because the tree-shaking.

**Comment by jbucaran** [comment]
> Another thing is that it does not activates tree-shaking on commonjs modules 

It does! Which is why that's an option I'm considering too to have everything and still remain cjs.

Notice that the above assumes you understand that our CDN is currently distributed as an iife.

**Comment by tunnckoCore** [comment]
> It does!

No, it doesn't. I'm using it for months and read tons of things - it was said somewhere, and makes perfect sense - even if we use that exact plugins. At least you need to provide `module` field for users to be able to use the tree shaking and to not include everything from hyperapp while they develop their apps.

First, you don't have `module` field. And second you don't export hyperapp as ES module, but just IIFE/UMD bundle.

**Comment by jbucaran** [comment]
@tunnckoCore  I decided to go with ES6 modules because it's less problematic for **end users**, and I'll be cutting a new release soon. 

> No, it doesn't. I'm using it for months and read tons of things...

Seems like you've misunderstood me, so I've created an example to show you or anyone else.

As you will see below, bundle.js contains only the exports that were actually used. 

#### Example

* index.js
```js
module.exports.a = 1
module.exports.b = 2
module.exports.c = 3
```

* app.js
```js
import { a, b } from "./"
console.log(a + b)
```

* rollup.config.js
```js
import cjs from "rollup-plugin-commonjs"
export default {
    plugins: [ cjs() ]
}
```

* bundle.js
```js
(function () {
'use strict';

var a = 1;
var b = 2;

console.log(a + b);

}());
```

* package.json

```json
{
  "name": "cjs-rollup-tree-shake-example",
  "version": "*",
  "scripts": {
    "build": "rollup -cf iife -n app -i app.js -o bundle.js"
  },
  "devDependencies": {
    "rollup": "^0.41.4",
    "rollup-plugin-commonjs": "^7.0.0"
  }
}
```

### so

Using the same tweak, we could keep hyperapp commonjs and still get treeshake work with rollup by changing index.js

```js
module.exports.h= require("./h")
module.exports.app= require("./app")
module.exports.router= require("./router")
```

**Comment by jbucaran** [comment]
>  Btw IIFE isn't good choice in anyway.

Depends on what you're doing. When and if hyperapp import/export semantics are changed to use ES6 modules, we'll need to distribute a UMD module so that `pkg.main` can be set to `dist/hyperapp.js` and hyperapp can work in node.

**Comment by tunnckoCore** [comment]
Anyway. Nevermind. I'm just too aggressive and angry today, I'll just give up. :)

**Comment by jbucaran** [comment]
@tunnckoCore It's cool. I've forgotten to add also, one I don't want to just go ahead and tweak index.js is because of this:

<img width="711" alt="screen shot 2017-02-15 at 3 53 21" src="https://cloud.githubusercontent.com/assets/56996/22943911/76dba5d6-f332-11e6-9af8-279df128c6a4.png">

And dist/hyperapp.js ends up with a `default` export.

```
/* ...
 h, app and router source code ...
*/

var h = h$1;
var app = app$1;
var router = router$1;

var index = {
	h: h,
	app: app,
	router: router
};

exports['default'] = index;
exports.h = h;
exports.app = app;
exports.router = router;

}((this.hyperapp = this.hyperapp || {})));
//# sourceMappingURL=hyperapp.js.map
```

**Comment by tunnckoCore** [comment]
Doh. :D You maybe just don't understand how all works and doing it wrong, don't know or i'm too awful today, actually just last couple of hours. I'm kinda lunatic, maybe it is because the moon, don't know :rofl: Don't get me offensive, please :D It is just short time period and tomorrow i can be super active with hundreds of commits. But anyway. offtopic end.

The UMD users won't have that `default` thing. They will just have `hyperapp` as usual. ES Users don't care they will have that `default`, because bundlers works that way. CJS users won't have that too - see the [commonjs bundle of RandomOrgJS](https://github.com/tunnckoCore/randomorg-js/blob/master/dist/randomorg.common.js#L75). So when users just `require` through node they are okey, dont need to do `require('randomorg-js).default` as seen in some places around github. If users wanna use the ES bundle they need just `from 'randomorg-js'` again, no `default` is needed anywhere. If browser users want use it they just use what I defined in the `moduleName: 'foobar'` - no need for `default`.

**Comment by tunnckoCore** [comment]
> exports['default'] = index;
exports.h = h;
exports.app = app;
exports.router = router;

Don't know how you get this, but it is wrong. And won't/shoudn't be that way. Try this.

**src/index.js**

```js
import router from './router'
import app from './app'
import h from './h'

export default {
  router,
  app,
  h
}
```

**src/app.js**

```js
export default function app (name) {
  return 123 + 'app'
}
```

**src/h.js**

```js
export default function h (tag, props, children) {
  return 123 + 'virtual dom'
}
```

**src/router.js**

```js
export default function router (routes, opts) {
  return 'the router' + 333
}
```

**rollup.config.js** - builds every kind of bundle - UMD, CJS, ES, IIFE - so everyone should be happy.

```js
import buble from 'rollup-plugin-buble'
import uglify from 'rollup-plugin-uglify'
import zopfli from 'rollup-plugin-zopfli'

let config = {
  entry: 'src/index.js'
}

if (process.env.BROWSER) {
  config = Object.assign(config, {
    // dest: 'dist/hyperapp.min.js',
    // format: 'umd',
    moduleName: 'hyperapp',
    useStrict: false,
    sourceMap: true,
    plugins: [
      buble({
        target: {
          ie: '10',
          edge: '12',
          safari: '8',
          chrome: '48',
          firefox: '44'
        }
      }),
      uglify({ compress: { warnings: false } }),
      zopfli({ options: { numiterations: 1000 } })
    ],
    targets: [
      { dest: 'dist/hyperapp.iife.js', format: 'iife' },
      { dest: 'dist/hyperapp.umd.js', format: 'umd' }
    ]
  })
} else {
  config = Object.assign(config, {
    plugins: [
      // just for sure
      buble({
        target: { node: '4' }
      })
    ],
    targets: [
      { dest: 'dist/hyperapp.common.js', format: 'cjs' },
      { dest: 'dist/hyperapp.es.js', format: 'es' }
    ]
  })
}

export default config
```

1. Uses Buble, intentionally, just for sure, cost nothing.
2. Your code can still be written in old ES, it just enough to use ES import/export.

**dist/hyperapp.common.js** - set to `pkg.main` for normal users and browserify

```js
'use strict';

function router (routes, opts) {
  return 'the router' + 333
}

function app (name) {
  return 123 + 'app'
}

function h (tag, props, children) {
  return 123 + 'virtual dom'
}

var index = {
  router,
  app,
  h
};

module.exports = index;
```

> Normal oldschool users just use as any other module - no `default` just `require('hyperapp')`.

**dist/hyperapp.es.js** - set to `pkg.module`

```js
function router (routes, opts) {
  return 'the router' + 333
}

function app (name) {
  return 123 + 'app'
}

function h (tag, props, children) {
  return 123 + 'virtual dom'
}

var index = {
  router,
  app,
  h
};

export default index;
```

> ES users don't care of that `export default`. They just can `import { h } from 'hyperapp'`. Tree-shaking enabled - they will just get the `h` while they build/bundle their app or whatever.

**dist/hyperapp.umd.js** - dont set it to `pkg.browser`, it is just for using through `unpkg.com`, force users to get it explicitly by writing `unpkg.com/hyperapp/dist/hyperapp.umd.js`, because just `unpkg.com/hyperapp` would drive them to the CJS bundle from the `pkg.main` (because we intentionally wouldn't set the `pkg.browser` field

```js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? module.exports = factory() :
  typeof define === 'function' && define.amd ? define(factory) :
  (global.hyperapp = factory());
}(this, (function () {

function router (routes, opts) {
  return 'the router' + 333
}

function app (name) {
  return 123 + 'app'
}

function h (tag, props, children) {
  return 123 + 'virtual dom'
}

var index = {
  router: router,
  app: app,
  h: h
};

return index;

})));
//# sourceMappingURL=hyperapp.umd.js.map
```

They still can just use `hyperapp` global, without problems and there's no need for `hyperapp['default']`

```html
<script src="https://unpkg.com/hyperapp/dist/hyperapp.umd.js"></script>
<script>
  const { h, app } = hyperapp
</script>
```

And same is for the `IIFE` bundle.
